drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,1931 @@
+{
+	/* For safe, EP (re-)enable from clean state */
+	numaker_usbd_ep_abort(ep_cur);
+	numaker_usbd_ep_clear_stall_n_data_toggle(ep_cur);
+	numaker_usbd_ep_fifo_reset(ep_cur);
+
+	/* Enable EP to IN/OUT */
+	ep_cur->usbd_hw_ep_cfg &= ~USBD_CFG_STATE_Msk;
+	if (USB_EP_DIR_IS_IN(ep_cur->ep_addr)) {
+		ep_cur->usbd_hw_ep_cfg |= USBD_CFG_EPMODE_IN;
+	} else {
+		ep_cur->usbd_hw_ep_cfg |= USBD_CFG_EPMODE_OUT;
+	}
+
+	USBD_CONFIG_EP(ep_cur->usbd_hw_ep_hndl, ep_cur->usbd_hw_ep_cfg);
+
+	/* For USBD, no separate EP interrupt control */
+}
+
+/* Disable EP */
+static void numaker_usbd_ep_disable(struct numaker_usbd_ep *ep_cur)
+{
+	/* For USBD, no separate EP interrupt control */
+
+	/* Disable EP */
+	ep_cur->usbd_hw_ep_cfg =
+		(ep_cur->usbd_hw_ep_cfg & ~USBD_CFG_STATE_Msk) | USBD_CFG_EPMODE_DISABLE;
+	USBD_CONFIG_EP(ep_cur->usbd_hw_ep_hndl, ep_cur->usbd_hw_ep_cfg);
+}
+
+/* Start EP data transaction */
+static void numaker_usbd_ep_trigger(struct numaker_usbd_ep *ep_cur, uint32_t len)
+{
+	struct usb_dc_numaker_device *dev = ep_cur->dev;
+	const struct usb_dc_numaker_config *config = dev->config;
+
+	USBD_SET_PAYLOAD_LEN(ep_cur->usbd_hw_ep_hndl, len);
+
+#if defined(CONFIG_USB_DC_NUMAKER_WORKAROUND_DISALLOW_ISO_IN_OUT_SAME_NUM)
+	if (ep_cur->usbd_hw_ep_cfg & USBD_CFG_TYPE_ISO) {
+		/* Enable forcing ISO IN Tx on ISO IN token arrival (even without above
+		 * USBD_SET_PAYLOAD_LEN()), to not get blocked by ISO OUT
+		 */
+		*((volatile uint32_t *)(config->usbd_base->RESERVE0 + 0)) &= 0xFFFFFFFD;
+	}
+#endif
+}
+
+/* Allocate EP context */
+static struct numaker_usbd_ep *numaker_usbd_ep_mgmt_alloc_ep(struct usb_dc_numaker_device *dev)
+{
+	const struct usb_dc_numaker_config *config = dev->config;
+	struct usb_dc_numaker_data *data = dev->data;
+	struct numaker_usbd_ep_mgmt *ep_mgmt = &data->ep_mgmt;
+	struct numaker_usbd_ep *ep_cur = NULL;
+
+	if (ep_mgmt->ep_idx < config->num_bidir_endpoints) {
+		ep_cur = ep_mgmt->ep_pool + ep_mgmt->ep_idx;
+		ep_mgmt->ep_idx++;
+
+		__ASSERT_NO_MSG(!ep_cur->valid);
+
+		/* Indicate this EP context is allocated */
+		ep_cur->valid = true;
+	}
+
+	return ep_cur;
+}
+
+/* Allocate DMA buffer
+ *
+ * Return -ENOMEM  on OOM error, or 0 on success with DMA buffer base/size (rounded up) allocated
+ */
+static int numaker_usbd_ep_mgmt_alloc_dmabuf(struct usb_dc_numaker_device *dev, uint32_t size,
+					     uint32_t *dmabuf_base_p, uint32_t *dmabuf_size_p)
+{
+	const struct usb_dc_numaker_config *config = dev->config;
+	struct usb_dc_numaker_data *data = dev->data;
+	struct numaker_usbd_ep_mgmt *ep_mgmt = &data->ep_mgmt;
+
+	__ASSERT_NO_MSG(dmabuf_base_p);
+	__ASSERT_NO_MSG(dmabuf_size_p);
+
+	/* Required to be 8-byte aligned */
+	size = ROUND_UP(size, 8);
+
+	ep_mgmt->dmabuf_pos += size;
+	if (ep_mgmt->dmabuf_pos > config->dmabuf_size) {
+		ep_mgmt->dmabuf_pos -= size;
+		return -ENOMEM; /* OOM error */
+	}
+
+	*dmabuf_base_p = ep_mgmt->dmabuf_pos - size;
+	*dmabuf_size_p = size;
+	return 0; /* Success */
```

Comment:
nit: redundant comment, I think everyone here knows what a `return 0;` at the end of the function means :-)

Reply:
dropped also